src/wcl/filepath.h

```diff
@@ -100,12 +102,47 @@ class filepath_range {
   filepath_iterator end() const { return filepath_iterator(str, str.size()); }
 };
 
-filepath_range make_filepath_range(const std::string& str) { return filepath_range(str); }
+inline filepath_range make_filepath_range(const std::string& str) { return filepath_range(str); }
 
-filepath_range make_filepath_range(std::string&& str) { return filepath_range(std::move(str)); }
+inline filepath_range make_filepath_range(std::string&& str) {
+  return filepath_range(std::move(str));
+}
 
-filepath_range_ref make_filepath_range_ref(const std::string& str) {
+inline filepath_range_ref make_filepath_range_ref(const std::string& str) {
   return filepath_range_ref(str);
 }
 
+inline std::string join_paths(std::string a, const std::string& b) {
+  if (a.back() == '/') {
```

Comment:
can you unnest this function with some early returns?

Reply:
So I did it and it looks like this:

```
inline std::string join_paths(std::string a, const std::string& b) {
  if (a.back() == '/') {
    if (b.front() == '/') {
      a.insert(a.end(), b.begin() + 1, b.end());
      return a;
    }
    a += b;
    return a;
  }
  if (b.front() == '/') {
    a += b;
    return a;
  }
  a += '/';
  a += b;

  return a;
}
```

but that's actually less understandable to me. There's no real preferred path here which is rare. Do you still want it to be return-ified?

Reply:
hmm maybe this is clearer?

```
inline std::string join_paths(std::string a, const std::string& b) {
  if (a.back() == '/' && b.front() == '/') {
    a.insert(a.end(), b.begin() + 1, b.end());
    return a;
  }
  if (a.back() != '/' && b.front() == '/')  {
    a += b;
    return a;
  }
  a += '/';
  a += b;

  return a;
}
```

or even

```
inline std::string join_paths(std::string a, const std::string& b) {
  if (b.front() == '/') {
      if(a.back() == '/') {
          a.insert(a.end(), b.begin() + 1, b.end());
          return a;
      }
     a += b;
     return a;
  }
  
  a += '/';
  a += b;

  return a;
}
```

Yeah, I really prefer this last one

Reply:
Neither of those work correctly because if b.front() != '/' but a.back() == '/' then you wind up appending an extra '/'

I think there's no real getting around having 4 cases here frankly.

Reply:
ok got something better!